### PR TITLE
[tx_pool] Invalidate txs with gasPrice < gasPriceMinimumFloor

### LINKED ISSUE #1768: Validate tx gasPrice against gasPriceMinimumFloor (prune pending txs)

### PR DESCRIPTION
**Pull request by gastonponti**
### Description

This fixes the issue that adds txs with gasPrice zero, which will not be executed but we keep for at lest 3 hours (default expiration time), and fill the txpool with peding txs

### Tested

Manually and add unit tests

### Related issues

- Fixes #1768 

### Backwards compatibility

Yes

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 35c5ef8cbcf3cdd02d62a63fe447246cf2d5fd2b
<details><summary>coverage:  44.5% of statements across all listed packages</summary><br>
<pre>
coverage:  54.1% of statements in consensus/istanbul
coverage:  44.9% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.4% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: 78d2c8d116
</details>

File: core/tx_pool_test.go [hunk 1]
```diff
@@ -400,6 +400,20 @@ func TestTransactionNegativeValue(t *testing.T) {
 	}
 }
 
+func TestTransactionZeroGasPrice(t *testing.T) {
+	t.Parallel()
+
+	pool, key := setupTxPool()
+	defer pool.Stop()
+
+	tx, _ := types.SignTx(types.NewTransaction(0, common.Address{}, big.NewInt(1), 200000, big.NewInt(0), nil, nil, nil, nil), types.HomesteadSigner{}, key)
```
Comment (piersy) [inline core/tx_pool_test.go:409]: If I try to send this tx using an e2e test, I get an error saying "replay protection is required", should we be using the EIP155Signer here ?

Comment (trianglesphere) [reply core/tx_pool_test.go:409]: Probably. We can also use `LatestSigner` to automatically keep up to date with changes in the signer.

**Review by trianglesphere — APPROVED** [review]
Nice find with the GPM floor.

**Comment by trianglesphere** [comment]
It looks like this is exercising a flaky codepath at startup whereby if the miner doesn't have a pending block, it fails on gas estimation.
Sample patch. I'm not sure if sending pending block to be the last block if there's no actual pending block makes sense
```
@@ -141,6 +141,9 @@ func (b *EthAPIBackend) StateAndHeaderByNumber(ctx context.Context, number rpc.B
        // Pending state is only known by the miner
        if number == rpc.PendingBlockNumber {
                block, state := b.eth.miner.Pending()
+               if block == nil && state == nil {
+                       return nil, nil, errors.New("No pending block")
+               }
                return state, block.Header(), nil
        }
```

**Pull request merged**